File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        n = len(nums)
        used = [False] * n

        def backtracking(curr):
            if len(curr) == n:
                res.append(curr[:])
                return
            
            for i in range(n):
                if used[i]:
                    continue

                used[i] = True
                curr.append(nums[i])

                backtracking(curr)

                curr.pop()
                used[i] = False
        
        backtracking([])

        return res

File: Data Structures & Algorithms/counting-bits/submission-1.cpp
class Solution {
public:
    vector<int> countBits(int n) {
        vector<int> res(n + 1);
        int sub = 1;

        for (int i = 1; i < n + 1; i++) {
            if (sub * 2 == i) {
                sub = i;
            }
            
            res[i] = res[i - sub] + 1;
        }

        return res;
    }
};